floors/rule_test.go

```diff
@@ -0,0 +1,776 @@
+package floors
+
+import (
+	"encoding/json"
+	"errors"
+	"testing"
+
+	"github.com/prebid/openrtb/v17/openrtb2"
+	"github.com/prebid/prebid-server/currency"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPrepareRuleCombinations(t *testing.T) {
+	testCases := []struct {
+		name string
+		in   []string
+		n    int
+		del  string
+		out  []string
+	}{
+		{
+			name: "Schema items, n = 1",
+			in:   []string{"A"},
+			n:    1,
+			del:  "|",
+			out: []string{
+				"a",
+				"*",
+			},
+		},
+		{
+			name: "Schema items, n = 2",
+			in:   []string{"A", "B"},
+			n:    2,
+			del:  "|",
+			out: []string{
+				"a|b",
+				"a|*",
+				"*|b",
+				"*|*",
+			},
+		},
+		{
+			name: "Schema items, n = 3",
+			in:   []string{"A", "B", "C"},
+			n:    3,
+			del:  "|",
+			out: []string{
+				"a|b|c",
+				"a|b|*",
+				"a|*|c",
+				"*|b|c",
+				"a|*|*",
+				"*|b|*",
+				"*|*|c",
+				"*|*|*",
+			},
+		},
+		{
+			name: "Schema items, n = 4",
+			in:   []string{"A", "B", "C", "D"},
+			n:    4,
+			del:  "|",
+			out: []string{
+				"a|b|c|d",
+				"a|b|c|*",
+				"a|b|*|d",
+				"a|*|c|d",
+				"*|b|c|d",
+				"a|b|*|*",
+				"a|*|c|*",
+				"a|*|*|d",
+				"*|b|c|*",
+				"*|b|*|d",
+				"*|*|c|d",
+				"a|*|*|*",
+				"*|b|*|*",
+				"*|*|c|*",
+				"*|*|*|d",
+				"*|*|*|*",
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			out := prepareRuleCombinations(tc.in, tc.n, tc.del)
```

Comment:
I believe the first parameter to this function is all of the desired rule keys for the impression which may include wildcards right? If so, I think we need to add some test cases here where `in` contains wildcards to ensure those scenarios work properly.

Reply:
added test cases to handle wildcards